game-core/src/main/java/games/strategy/net/IClientMessenger.java

```diff
@@ -34,4 +35,13 @@
   @SuppressWarnings("unused")
   @Nullable
   ApiKey getApiKey();
+
+  default ApiKey getApiKeyOrThrow() {
+    return Optional.ofNullable(getApiKey())
+        .orElseThrow(
+            () ->
+                new IllegalStateException(
+                    "Unexpected missing api key, programmer error. "
+                        + "Likely caused by trying to access API key for a non-lobby connection."));
```

Comment:
Yeah, the code smell is definitely there. It was less so before because messengers is so generic and you can basically call any method you want via RMI. Having messengers was basically like having no interface at all almost.

The root of the smell is that the IClientMessenger is used for all client connection types. Not all client connection types will support an API key. Using `UnsupportedOperationException` makes sense

Reply:
Grin, also check out the other APIs in this interface, most are to control headless game hosts.. The getApiKey here, I hope will be transitory and eventually would be removed.